File: src/org/puder/trs80/keyboard/Key.java
/*
 * Copyright 2012-2013, Arno Puder
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.puder.trs80.keyboard;

import org.puder.trs80.Hardware;
import org.puder.trs80.R;
import org.puder.trs80.TRS80Application;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Paint.Align;
import android.graphics.Paint.Style;
import android.graphics.RectF;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup.MarginLayoutParams;

/**
 * The emulator features different keyboard layouts (original, compact, etc).
 * The XML layout resources can be found in res/layout/keyboard_*.xml. Class Key
 * implements the behavior of one key of the keyboard. Class Key is a custom
 * Android widget that is referenced from the aforementioned XML layout files.
 * Whenever the user 'clicks' on a key, class Key uses memory mapped IO to poke
 * a bit into the emulated RAM of the TRS-80. See the onTouch() method further
 * below. The custom XML attribute keyboard:id is used to cross-reference a key
 * map via the KeyboardManager (e.g., the Model 3 key map is stored in
 * keymap_model3.xml and loaded by KeyboardManager). This file contains the
 * memory addresses and bit mask for each key. The emulator runs in a separate
 * thread and will 'see' that a bit is flipped at a certain memory address and
 * will interpret this as a key-press. See the following link for the
 * memory-mapped IO address of the Model 3 keyboard:
 * 
 * http://www.trs-80.com/wordpress/zaps-patches-pokes-tips/internals/#keyboard13
 */
public class Key extends View {

    final static int        TK_0                = 0;
    final static int        TK_1                = 1;
    final static int        TK_2                = 2;
    final static int        TK_3                = 3;
    final static int        TK_4                = 4;
    final static int        TK_5                = 5;
    final static int        TK_6                = 6;
    final static int        TK_7                = 7;
    final static int        TK_8                = 8;
    final static int        TK_9                = 9;
    final static int        TK_A                = 10;
    final static int        TK_B                = 11;
    final static int        TK_C                = 12;
    final static int        TK_D                = 13;
    final static int        TK_E                = 14;
    final static int        TK_F                = 15;
    final static int        TK_G                = 16;
    final static int        TK_H                = 17;
    final static int        TK_I                = 18;
    final static int        TK_J                = 19;
    final static int        TK_K                = 20;
    final static int        TK_L                = 21;
    final static int        TK_M                = 22;
    final static int        TK_N                = 23;
    final static int        TK_O                = 24;
    final static int        TK_P                = 25;
    final static int        TK_Q                = 26;
    final static int        TK_R                = 27;
    final static int        TK_S                = 28;
    final static int        TK_T                = 29;
    final static int        TK_U                = 30;
    final static int        TK_V                = 31;
    final static int        TK_W                = 32;
    final static int        TK_X                = 33;
    final static int        TK_Y                = 34;
    final static int        TK_Z                = 35;
    final static int        TK_COMMA            = 36;
    final static int        TK_DOT              = 37;
    final static int        TK_SLASH            = 38;
    final static int        TK_SPACE            = 39;
    final static int        TK_ADD              = 40;
    final static int        TK_HASH             = 41;
    final static int        TK_BR_OPEN          = 42;
    final static int        TK_BR_CLOSE         = 43;
    final static int        TK_ASTERIX          = 44;
    final static int        TK_DOLLAR           = 45;
    final static int        TK_QUESTION         = 46;
    final static int        TK_LT               = 47;
    final static int        TK_GT               = 48;
    final static int        TK_EQUAL            = 49;
    final static int        TK_PERCENT          = 50;
    final static int        TK_APOS             = 51;
    final static int        TK_EXCLAMATION_MARK = 52;
    final static int        TK_AMP              = 53;
    final static int        TK_QUOT             = 54;
    final static int        TK_SEMICOLON        = 55;
    final static int        TK_ENTER            = 56;
    final static int        TK_CLEAR            = 57;
    final static int        TK_CLEAR_SHORT      = 58;
    final static int        TK_SHIFT_LEFT       = 59;
    final static int        TK_SHIFT_RIGHT      = 60;
    final static int        TK_COLON            = 61;
    final static int        TK_MINUS            = 62;
    final static int        TK_BREAK            = 63;
    final static int        TK_BREAK_SHORT      = 64;
    final static int        TK_UP               = 65;
    final static int        TK_AT               = 66;
    final static int        TK_LEFT             = 67;
    final static int        TK_RIGHT            = 68;
    final static int        TK_DOWN             = 69;
    final static int        TK_ALT              = 70;

    private int             idNormal;
    private int             idShifted;

    private String          labelNormal;
    private String          labelShifted;
    private boolean         isShifted;
    private boolean         isPressed;
    private boolean         isShiftKey;
    private boolean         isAltKey;

    private View            keyboardView1;
    private View            keyboardView2;

    private int             size;
    private Paint           paint;
    private RectF           rect;

    private KeyboardManager keyboard;

    private int             keyWidth;
    private int             keyHeight;
    private int             keyMargin;

    private int             posX                = -1;
    private int             posY                = -1;

    public Key(Context context, AttributeSet attrs) {
        super(context, attrs);

        this.setBackgroundResource(R.drawable.key_background);

        Hardware h = TRS80Application.getHardware();
        keyWidth = h.getKeyWidth();
        keyHeight = h.getKeyHeight();
        keyMargin = h.getKeyMargin();

        keyboard = TRS80Application.getKeyboardManager();
        TypedArray ta = context.obtainStyledAttributes(attrs, R.styleable.Keyboard, 0, 0);
        idNormal = ta.getInt(R.styleable.Keyboard_id, -1);
        idShifted = ta.getInt(R.styleable.Keyboard_idShifted, -1);
        if (idNormal == TK_ALT) {
            labelNormal = "Alt";
        } else {
            KeyMap keyMap = keyboard.getKeyMap(idNormal);
            labelNormal = keyMap.label;
        }
        size = ta.getInteger(R.styleable.Keyboard_size, 1);
        ta.recycle();

        if (idShifted != -1) {
            KeyMap keyMap = keyboard.getKeyMap(idShifted);
            labelShifted = keyMap.label;
            keyboard.addShiftableKey(this);
        }
        isShiftKey = idNormal == TK_SHIFT_LEFT || idNormal == TK_SHIFT_RIGHT;
        if (isShiftKey) {
            labelShifted = labelNormal;
            keyboard.addShiftableKey(this);
        }
        isShifted = false;
        isPressed = false;
        isAltKey = idNormal == TK_ALT;
        paint = new Paint();
        paint.setTypeface(TRS80Application.getTypeface());
        paint.setAntiAlias(true);
        float textSizeScale = labelNormal.length() > 1 ? 0.4f : 0.6f;
        paint.setTextSize(keyHeight * textSizeScale);

        rect = new RectF();
        this.setOnTouchListener(new OnTouchListener() {

            @Override
            public boolean onTouch(View view, MotionEvent event) {
                int action = event.getAction() & MotionEvent.ACTION_MASK;
                if (action == MotionEvent.ACTION_DOWN) {
                    isPressed = true;
                    invalidate();
                }
                if (action == MotionEvent.ACTION_UP) {
                    isPressed = false;
                    invalidate();
                }
                if (isAltKey) {
                    if (action == MotionEvent.ACTION_UP) {
                        switchKeyboard();
                    }
                    return true;
                }
                if (action == MotionEvent.ACTION_DOWN) {
                    if (isShifted && idShifted != -1) {
                        keyboard.keyDown(idShifted);
                    } else {
                        keyboard.keyDown(idNormal);
                    }
                }
                if (action == MotionEvent.ACTION_UP) {
                    if (isShiftKey) {
                        if (!isShifted) {
                            keyboard.shiftKeys();
                        } else {
                            keyboard.unshiftKeys();
                        }
                    } else {
                        if (isShifted && idShifted != -1) {
                            keyboard.keyUp(idShifted);
                        } else {
                            keyboard.keyUp(idNormal);
                        }
                        keyboard.unshiftKeys();
                    }
                }
                return true;
            }
        });
    }

    @Override
    public void onDraw(Canvas canvas) {
        if (isShifted) {
            paint.setColor(Color.WHITE);
            paint.setAlpha(70);
            paint.setStyle(Style.FILL);
            canvas.drawRoundRect(rect, 10, 10, paint);
        }

        if (isPressed) {
            paint.setColor(Color.WHITE);
            paint.setAlpha(95);
            paint.setStyle(Style.FILL);
            canvas.drawRoundRect(rect, 10, 10, paint);
        }

        paint.setColor(Color.GRAY);
        paint.setAlpha(130);
        paint.setStyle(Style.STROKE);
        paint.setStrokeWidth(3);
        canvas.drawRoundRect(rect, 10, 10, paint);

        paint.setColor(Color.WHITE);
        paint.setAlpha(110);
        paint.setStyle(Style.FILL);
        paint.setStrokeWidth(1);
        paint.setTextAlign(Align.CENTER);
        int xPos = (int) (rect.right / 2);
        int yPos = (int) ((rect.bottom / 2) - ((paint.descent() + paint.ascent()) / 2));
        canvas.drawText(isShifted ? labelShifted : labelNormal, xPos, yPos, paint);
    }

    @Override
    protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
        super.onLayout(changed, left, top, right, bottom);

        if (isAltKey) {
            // We can only do this here to ensure that this view has already
            // been
            // added to the view hierarchy
            keyboardView1 = this.getRootView().findViewById(R.id.keyboard_view_1);
            keyboardView2 = this.getRootView().findViewById(R.id.keyboard_view_2);
        }

        MarginLayoutParams params = (MarginLayoutParams) this.getLayoutParams();
        if (posX != -1 && posY != -1) {
            params.setMargins(posX, posY, 0, 0);
        } else {
            params.setMargins(keyMargin, keyMargin, keyMargin, keyMargin);
        }
        this.setLayoutParams(params);
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        int width;
        int height;
        if ((widthMeasureSpec & MeasureSpec.EXACTLY) != 0) {
            // Layout specifies specific size. Use that.
            width = widthMeasureSpec & ~MeasureSpec.EXACTLY;
        } else {
            width = keyWidth * size;
        }
        if ((heightMeasureSpec & MeasureSpec.EXACTLY) != 0) {
            // Layout specifies specific size. Use that.
            height = heightMeasureSpec & ~MeasureSpec.EXACTLY;
        } else {
            height = keyHeight;
        }
        setMeasuredDimension(width | MeasureSpec.EXACTLY, height | MeasureSpec.EXACTLY);
        rect.set(1, 1, width - 1, height - 1);
    }

    public void setPosition(int x, int y) {
        posX = x;
        posY = y;
    }

    public void shift() {
        if (!isShifted) {
            isShifted = true;
            invalidate();
        }
    }

    public void unshift() {
        if (isShifted) {
            isShifted = false;
            invalidate();
        }
    }

    public void switchKeyboard() {
        if (keyboardView1.getVisibility() == View.GONE) {
            keyboardView1.setVisibility(View.VISIBLE);
            keyboardView2.setVisibility(View.GONE);
        } else {
            keyboardView1.setVisibility(View.GONE);
            keyboardView2.setVisibility(View.VISIBLE);
        }
    }
}

File: app/src/main/java/org/puder/trs80/keyboard/Key.kt
/*
 * Copyright 2012-2013, Arno Puder
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.puder.trs80.keyboard

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.RectF
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import android.view.View.MeasureSpec
import android.view.ViewGroup
import org.puder.trs80.EmulatorActivity
import org.puder.trs80.R
import org.puder.trs80.TypefaceCache
import kotlin.math.min

/**
 * A single key of the on-screen TRS-80 keyboard.
 *
 * Keys are inflated from the keyboard layouts, which supply the key id, the shifted key id and
 * the width of the key through the `Keyboard` styleable. The hosting context must be an
 * [EmulatorActivity], since the key takes its dimensions and its [KeyboardManager] from it.
 */
@SuppressLint("ClickableViewAccessibility", "CustomViewStyleable")
class Key @JvmOverloads constructor(
        context: Context,
        attrs: AttributeSet? = null,
        defStyle: Int = 0) : View(context, attrs, defStyle) {

    private val keyboard: KeyboardManager
    private val keyWidth: Int
    private val keyHeight: Int
    private val keyMargin: Int

    private val idNormal: Int
    private val idShifted: Int
    private val size: Int
    private val labelNormal: String
    private val labelShifted: String
    private val isShiftKey: Boolean
    private val isAltKey: Boolean

    private val paint = Paint()
    private val rect = RectF()

    private var isShifted = false
    private var isKeyPressed = false
    private var posX = -1
    private var posY = -1

    /** The two halves of the compact keyboard, resolved on first use, as the original did. */
    private val keyboardView1: View by lazy(LazyThreadSafetyMode.NONE) {
        rootView.findViewById(R.id.keyboard_view_1)
    }
    private val keyboardView2: View by lazy(LazyThreadSafetyMode.NONE) {
        rootView.findViewById(R.id.keyboard_view_2)
    }

    init {
        setBackgroundResource(R.drawable.key_background)

        val emulator = context as EmulatorActivity
        keyboard = emulator.keyboardManager
        keyWidth = emulator.keyWidth
        keyHeight = emulator.keyHeight
        keyMargin = emulator.keyMargin

        val ta = context.obtainStyledAttributes(attrs, R.styleable.Keyboard, 0, 0)
        idNormal = ta.getInt(R.styleable.Keyboard_id, -1)
        idShifted = ta.getInt(R.styleable.Keyboard_idShifted, -1)
        labelNormal = if (idNormal == TK_ALT) "Alt" else keyboard.getKeyMap(idNormal).label
        size = ta.getInteger(R.styleable.Keyboard_size, 1)
        ta.recycle()

        isShiftKey = idNormal == TK_SHIFT_LEFT || idNormal == TK_SHIFT_RIGHT
        isAltKey = idNormal == TK_ALT

        var shiftedLabel: String? = null
        if (idShifted != -1) {
            shiftedLabel = keyboard.getKeyMap(idShifted).label
            keyboard.addShiftableKey(this)
        }
        if (isShiftKey) {
            // A shift key keeps its own label when shifted.
            shiftedLabel = labelNormal
            keyboard.addShiftableKey(this)
        }
        labelShifted = shiftedLabel ?: labelNormal

        paint.typeface = TypefaceCache.getTypeface(FONT_PATH, context)
        paint.isAntiAlias = true
        paint.textSize = keyHeight * (if (labelNormal.length > 1) 0.4f else 0.6f)

        setOnTouchListener { _, event ->
            handleTouch(event.action and MotionEvent.ACTION_MASK)
            true
        }
    }

    /** Places the key at an absolute offset instead of using the uniform key margin. */
    fun setPosition(x: Int, y: Int) {
        posX = x
        posY = y
    }

    /** Switches the key to its shifted label, if it is not already showing it. */
    fun shift() {
        if (!isShifted) {
            isShifted = true
            invalidate()
        }
    }

    /** Switches the key back to its normal label, if it is not already showing it. */
    fun unshift() {
        if (isShifted) {
            isShifted = false
            invalidate()
        }
    }

    /** Toggles between the two halves of the compact keyboard layout. */
    fun switchKeyboard() {
        if (keyboardView1.visibility == View.GONE) {
            keyboardView1.visibility = View.VISIBLE
            keyboardView2.visibility = View.GONE
        } else {
            keyboardView1.visibility = View.GONE
            keyboardView2.visibility = View.VISIBLE
        }
    }

    override fun onDraw(canvas: Canvas) {
        if (isShifted) {
            fillRoundRect(canvas, SHIFTED_ALPHA)
        }
        if (isKeyPressed) {
            fillRoundRect(canvas, PRESSED_ALPHA)
        }

        paint.color = Color.GRAY
        paint.alpha = 130
        paint.style = Paint.Style.STROKE
        paint.strokeWidth = 3f
        canvas.drawRoundRect(rect, CORNER_RADIUS, CORNER_RADIUS, paint)

        paint.color = resources.getColor(R.color.key_color, null)
        paint.alpha = 110
        paint.style = Paint.Style.FILL
        paint.strokeWidth = 1f
        paint.textAlign = Paint.Align.CENTER
        // Both coordinates are deliberately truncated to whole pixels.
        val xPos = (rect.right / 2).toInt().toFloat()
        val yPos = (rect.bottom / 2 - (paint.descent() + paint.ascent()) / 2).toInt().toFloat()
        canvas.drawText(if (isShifted) labelShifted else labelNormal, xPos, yPos, paint)
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val desiredWidth = keyWidth * size
        val desiredHeight = keyHeight

        val width = when (MeasureSpec.getMode(widthMeasureSpec)) {
            MeasureSpec.EXACTLY -> MeasureSpec.getSize(widthMeasureSpec)
            MeasureSpec.AT_MOST -> min(desiredWidth, MeasureSpec.getSize(widthMeasureSpec))
            else -> desiredWidth
        }
        val height = when (MeasureSpec.getMode(heightMeasureSpec)) {
            MeasureSpec.EXACTLY -> MeasureSpec.getSize(heightMeasureSpec)
            MeasureSpec.AT_MOST -> min(desiredHeight, MeasureSpec.getSize(heightMeasureSpec))
            else -> desiredHeight
        }

        setMeasuredDimension(width, height)
        rect.set(1f, 1f, (width - 1).toFloat(), (height - 1).toFloat())

        val params = layoutParams as ViewGroup.MarginLayoutParams
        if (posX != -1 && posY != -1) {
            params.setMargins(posX, posY, 0, 0)
        } else {
            params.setMargins(keyMargin, keyMargin, keyMargin, keyMargin)
        }
        layoutParams = params
    }

    /**
     * Turns a masked [MotionEvent] action into emulator key events. Note that a normal key sends
     * its key-down on touch-down but a shift key toggles on touch-up, since it has to latch.
     */
    private fun handleTouch(action: Int) {
        when (action) {
            MotionEvent.ACTION_DOWN -> {
                isKeyPressed = true
                invalidate()
            }

            MotionEvent.ACTION_UP -> {
                isKeyPressed = false
                invalidate()
            }
        }

        if (isAltKey) {
            if (action == MotionEvent.ACTION_UP) {
                switchKeyboard()
            }
            return
        }

        if (action == MotionEvent.ACTION_DOWN && !isShiftKey) {
            keyboard.keyDown(activeId())
        }

        if (action != MotionEvent.ACTION_UP) {
            return
        }

        if (isShiftKey) {
            if (isShifted) {
                keyboard.unshiftKeys()
                keyboard.keyUp(idNormal)
            } else {
                keyboard.shiftKeys(idNormal)
                keyboard.keyDown(idNormal)
            }
            return
        }

        keyboard.keyUp(activeId())
        // Releasing any other key also releases a latched shift key.
        when (val shiftKey = keyboard.pressedShiftKey) {
            TK_SHIFT_LEFT, TK_SHIFT_RIGHT -> {
                keyboard.unshiftKeys()
                keyboard.keyUp(shiftKey)
            }
        }
    }

    /** @return The key id this key currently stands for, honouring the shift state. */
    private fun activeId(): Int = if (isShifted && idShifted != -1) idShifted else idNormal

    private fun fillRoundRect(canvas: Canvas, alpha: Int) {
        paint.color = Color.WHITE
        paint.alpha = alpha
        paint.style = Paint.Style.FILL
        canvas.drawRoundRect(rect, CORNER_RADIUS, CORNER_RADIUS, paint)
    }

    companion object {
        const val TK_NONE = -1
        const val TK_0 = 0
        const val TK_1 = 1
        const val TK_2 = 2
        const val TK_3 = 3
        const val TK_4 = 4
        const val TK_5 = 5
        const val TK_6 = 6
        const val TK_7 = 7
        const val TK_8 = 8
        const val TK_9 = 9
        const val TK_A = 10
        const val TK_B = 11
        const val TK_C = 12
        const val TK_D = 13
        const val TK_E = 14
        const val TK_F = 15
        const val TK_G = 16
        const val TK_H = 17
        const val TK_I = 18
        const val TK_J = 19
        const val TK_K = 20
        const val TK_L = 21
        const val TK_M = 22
        const val TK_N = 23
        const val TK_O = 24
        const val TK_P = 25
        const val TK_Q = 26
        const val TK_R = 27
        const val TK_S = 28
        const val TK_T = 29
        const val TK_U = 30
        const val TK_V = 31
        const val TK_W = 32
        const val TK_X = 33
        const val TK_Y = 34
        const val TK_Z = 35
        const val TK_COMMA = 36
        const val TK_DOT = 37
        const val TK_SLASH = 38
        const val TK_SPACE = 39
        const val TK_ADD = 40
        const val TK_HASH = 41
        const val TK_BR_OPEN = 42
        const val TK_BR_CLOSE = 43
        const val TK_ASTERIX = 44
        const val TK_DOLLAR = 45
        const val TK_QUESTION = 46
        const val TK_LT = 47
        const val TK_GT = 48
        const val TK_EQUAL = 49
        const val TK_PERCENT = 50
        const val TK_APOS = 51
        const val TK_EXCLAMATION_MARK = 52
        const val TK_AMP = 53
        const val TK_QUOT = 54
        const val TK_SEMICOLON = 55
        const val TK_ENTER = 56
        const val TK_CLEAR = 57
        const val TK_CLEAR_SHORT = 58
        const val TK_SHIFT_LEFT = 59
        const val TK_SHIFT_RIGHT = 60
        const val TK_COLON = 61
        const val TK_MINUS = 62
        const val TK_BREAK = 63
        const val TK_BREAK_SHORT = 64
        const val TK_UP = 65
        const val TK_AT = 66
        const val TK_LEFT = 67
        const val TK_RIGHT = 68
        const val TK_DOWN = 69
        const val TK_ALT = 70

        private const val FONT_PATH = "fonts/DejaVuSansMono.ttf"
        private const val CORNER_RADIUS = 10f
        private const val SHIFTED_ALPHA = 70
        private const val PRESSED_ALPHA = 95
    }
}
